[export] Fix torch.export() issues for server use cases.

zhxchen17 opened the pull request:
Test Plan: In D48788843

Differential Revision: D48811793

tugsbayasgalan commented on torch/_export/__init__.py:
```diff
@@ -433,24 +484,37 @@ def to_str_dict(sig_component: Dict[Any, Any]):
-    )
-    if len(preserve_module_call_signature) > 0:
-        exported_program = exported_program._transform(CollectTracepointsPass(module_call_signatures))
+    if len(range_constraints) > 0 or len(equality_constraints) > 0:
```
Do you know if we put the unbacked symint based ranges into above two lists?

zhxchen17 replied:
I think range_constraints is for both unbacked and backed symints?

tugsbayasgalan replied:
Yeah thanks for confirming!

tugsbayasgalan commented on torch/_export/passes/collect_tracepoints_pass.py:
```diff
@@ -18,6 +20,15 @@ def __init__(self, specs) -> None:
+                if isinstance(arg.meta.get('val'), torch.Tensor):
+                    return ArgumentSpec(kind=ArgumentKind.Tensor, value=arg.name)
+                else:
+                    raise RuntimeError("NYI")
```
NIT: Could you add a more descriptive error message?

tugsbayasgalan commented on torch/_export/__init__.py:
```diff
@@ -210,6 +210,36 @@ def _safe_to_skip_dynamo(gm: torch.fx.GraphModule):
+                assert path != root
+                assert root in path
+            # TODO(zhxchen17) normalize all the way down to dot strings.
+            node.meta["nn_module_stack"] = {'L__self__': (root, root_cls), **nn_module_stack}
```
NIT: Can you leave a comment on what you are changing in the spec?

tugsbayasgalan commented on torch/_export/__init__.py:
```diff
@@ -352,6 +382,26 @@ def export(
+        for dynamo_name, dynamo_param in gm_torch_level.named_parameters():
+            assert dynamo_name not in param_buffer_table
+            if id(dynamo_param) in param_lookup:
+                param_buffer_table[dynamo_name] = param_lookup[id(dynamo_param)]
```
Are you mapping dynamo param name to the original name?

zhxchen17 replied:
yes

tugsbayasgalan commented on torch/_export/__init__.py:
```diff
@@ -210,6 +210,36 @@ def _safe_to_skip_dynamo(gm: torch.fx.GraphModule):
 
+def _replace_param_buffer_names(param_buffer_table, sig):
+    replace = lambda x: param_buffer_table.get(x, x)
+    sig.parameters = pytree.tree_map(replace, sig.parameters)
```
OP way to replace :)

zhxchen17 committed:
[export] Fix torch.export() issues for server use cases. (#108275)

Summary: Pull Request resolved: https://github.com/pytorch/pytorch/pull/108275

Test Plan: In D48788843

Reviewed By: tugsbayasgalan

Differential Revision: D48811793

Pulled By: zhxchen17

fbshipit-source-id: ce3c7d2443b90d5e6045ca3f7e9be981d2ce7069

zhxchen17 commented:
@pytorchbot merge

desertfire commented:
@zhxchen17 , `python benchmarks/dynamo/huggingface.py --bfloat16 --accuracy --inference --device cuda --export --only  BertForMaskedLM` (and so are other benchmarks) fails after this PR. Reverting this PRK will make it pass. I guess we don't have any CI tasks for testing `--export` and `--export-aot-inductor` on the OSS benchmarks?

zhxchen17 commented:
> @zhxchen17 , `python benchmarks/dynamo/huggingface.py --bfloat16 --accuracy --inference --device cuda --export --only  BertForMaskedLM` (and so are other benchmarks) fails after this PR. Reverting this PRK will make it pass. I guess we don't have any CI tasks for testing `--export` and `--export-aot-inductor` on the OSS benchmarks?

Can I take a look and see if this forward fixable?

zhxchen17 commented:
> @zhxchen17 , `python benchmarks/dynamo/huggingface.py --bfloat16 --accuracy --inference --device cuda --export --only BertForMaskedLM` (and so are other benchmarks) fails after this PR. Reverting this PRK will make it pass. I guess we don't have any CI tasks for testing `--export` and `--export-aot-inductor` on the OSS benchmarks?

@desertfire yeah I didn't see oss test failures.

desertfire commented:
> > @zhxchen17 , `python benchmarks/dynamo/huggingface.py --bfloat16 --accuracy --inference --device cuda --export --only  BertForMaskedLM` (and so are other benchmarks) fails after this PR. Reverting this PRK will make it pass. I guess we don't have any CI tasks for testing `--export` and `--export-aot-inductor` on the OSS benchmarks?
> 
> Can I take a look and see if this forward fixable?

Go ahead, since it doesn't affect any CI at the moment, but if it takes more than a few hours, let's revert. And I will add CI tests after your fix or revert.

zhxchen17 commented:
@desertfire looks like there's a small bug, I'm fixing it in https://github.com/pytorch/pytorch/pull/108354.
